improc/simulator.py

```diff
@@ -0,0 +1,886 @@
+            # this convolution means that each new pixel is the SUM of all the pixels in the kernel
+            kernel = np.ones((oversampling, oversampling), dtype=float)
+            self.flux_top = scipy.signal.convolve(self.flux_top, kernel, mode='same')
+            self.flux_top = self.flux_top[oversampling//2::oversampling, oversampling//2::oversampling]
+            self.flux_top = self.flux_top[buffer[0]:-buffer[0], buffer[1]:-buffer[1]]
+
+    def add_atmosphere(self):
+        """
+        Add the effects of the atmosphere, namely the sky background and transmission.
+        """
+        self.flux_with_sky = self.flux_top * self.sky.transmission_instance + self.sky.background_instance
+
+    def add_camera(self):
+        """
+        Add the effects of the camera, namely the vignette.
+        """
+        self.flux_vignette = self.flux_with_sky * self.camera.vignette_map
+
+    def flux_to_electrons(self):
+        """
+        Calculate the number of electrons in each pixel,
+        accounting for the total QE and the pixel QE,
+        and adding the dark current.
+        """
+        self.electrons = self.flux_vignette * self.sensor.pixel_qe_map
+        self.electrons += self.sensor.dark_current * self.pars.exposure_time
+
+        # add saturation and bleeding:
+        # TODO: add bleeding before clipping
+        self.electrons[self.electrons > self.sensor.saturation_limit] = self.sensor.saturation_limit
+
+    def add_artefacts(self):
+        """
+        Add artefacts like cosmic rays, satellites, etc.
+        This should produce the final image.
+        """
+
+        for i in range(np.random.poisson(self.pars.cosmic_ray_number)):
+            self.add_cosmic_ray(self.average_counts)  # add in place
+
+        # TODO: satellites should be moved to the stage above atmosphere where stars and galaxies are added!
+        for i in range(np.random.poisson(self.pars.satellite_number)):
+            self.add_satellite(self.average_counts)  # add in place
+
+        # add more artefacts here...
+
+    def add_cosmic_ray(self, image):
+        """
+        Add a cosmic ray to the image.
+
+        Parameters
+        ----------
+        image: array
+            The image to add the cosmic ray to.
+            This will be modified in place.
+        """
+        pass
+
+    def add_satellite(self, image):
+        """
+        Add a satellite trail to the image.
+
+        Parameters
+        ----------
+        image: array
+            The image to add the satellite trail to.
+            This will be modified in place.
+        """
+        pass
+
+    def electrons_to_adu(self):
+        """
+        Convert the number of electrons in each pixel
+        to the number of ADU (analog to digital units)
+        that will be read out.
+        """
+        self.average_counts = self.electrons * self.sensor.pixel_gain_map
+        self.noise_var_map = self.average_counts + self.sensor.read_noise ** 2
```

Comment:
Poisson noise should be based on the number of electrons, not ADU.

Reply:
You are right! I knew it would be useful to have someone else go over the actual simulation code :)